calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,858 @@
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::RRC;
+use crate::ir::{self, GetAttributes};
+use std::collections::HashMap;
+
+#[derive(Default)]
+/// 1. loop through all control statements under "par" block to find # barriers
+///    needed and # members of each barrier
+/// 2. add all cells and groups needed
+/// 3. loop through all control statements, find the statements with @sync
+///    attribute and replace them with
+///     seq {
+///       <stmt>;
+///       incr_barrier_0_*;
+///       write_barrier_0_*;
+///       wait_*;
+///       restore_*;
+///     }
+///    or
+///     seq {
+///       <stmt>;
+///       incr_barrier_*_*;
+///       write_barrier_*_*;
+///       wait_*;
+///       wait_restore_*;
+///     }
+
+pub struct CompileSync;
+
+impl Named for CompileSync {
+    fn name() -> &'static str {
+        "compile-sync"
+    }
+
+    fn description() -> &'static str {
+        "Implement barriers for statements marked with @sync attribute"
+    }
+}
+
+fn count_barriers(s: &ir::Control, map: &mut HashMap<u64, u64>) -> () {
+    match s {
+        ir::Control::Seq(seq) => {
+            for stmt in seq.stmts.iter() {
+                if let Some(n) = stmt.get_attributes().unwrap().get("sync") {
+                    map.entry(*n).and_modify(|count| *count += 1).or_insert(1);
+                }
+            }
+        }
+        ir::Control::While(w) => {
+            count_barriers(&w.body, map);
+        }
+        ir::Control::If(i) => {
+            count_barriers(&i.tbranch, map);
+            count_barriers(&i.fbranch, map);
+        }
+        _ => {}
+    }
+}
+
+fn fmt_shared_name(comp: &'static str, idx: &u64) -> ir::Id {
+    format!("{}_{}", comp, idx).into()
+}
+
+fn fmt_ind_name(
+    comp: &'static str,
+    idx_barrier: &u64,
+    idx_member: &u64,
+) -> ir::Id {
+    format!("{}_{}_{}", comp, idx_barrier, idx_member).into()
+}
+
+fn add_shared_primitive(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    primitive: &str,
+    parameters: &Vec<u64>,
+    map: &mut HashMap<ir::Id, ir::Id>,
+) -> RRC<ir::Cell> {
+    let prim = builder.add_primitive(
+        fmt_shared_name(prefix, barrier_idx),
+        primitive,
+        parameters,
+    );
+    map.insert(
+        fmt_shared_name(prefix, barrier_idx),
+        prim.borrow().name().clone(),
+    );
+    prim
+}
+
+fn add_ind_primitive(
+    builder: &mut ir::Builder,
+    prefix: &'static str,
+    barrier_idx: &u64,
+    member_idx: &u64,
+    primitive: &str,
+    parameters: &Vec<u64>,
+    map: &mut HashMap<ir::Id, ir::Id>,
```

Comment:
This `map` should probably be a struct that wraps all of these methods and makes them semantically meaninful